### PR TITLE
msct_nurbs: fixes singular matrix error

### LINKED ISSUE #1863: numpy.linalg.linalg.LinAlgError: singular matrix
data: mgh-bay3_spine-generic_20171201/t2
~~~
Spinal Cord Toolbox (master/c1e4e9cefd485570916857e17f3703d728160f2a)
Running /Users/julien/code/sct/scripts/sct_process_segmentation.py -i t2_seg_manual.nii.gz -p csa -vert 2 -vertfile t1_seg_labeled_reg.nii.gz -ofolder csa

Check parameters:
.. segmentation file:             t2_seg_manual.nii.gz

Create temporary folder (/var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180613161703-PkT5TR)...

Copying input data to tmp folder and convert to nii...
sct_convert -i /Users/julien/code/spine_generic/mgh-bay3_spine-generic_20171201/t2/t2_seg_manual.nii.gz -o /var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180613161703-PkT5TR/segmentation.nii.gz # in /Users/julien/code/spine_generic/mgh-bay3_spine-generic_20171201/t2

Change orientation to RPI...
sct_image -i segmentation.nii.gz -setorient RPI -o segmentation_RPI.nii.gz # in /private/var/folders/6f/wy6ljmx9453cgth2qwv5l1s80000gn/T/sct-180613161703-PkT5TR

Open segmentation volume...

Get data dimensions...
  64 x 320 x 320

Smooth centerline/segmentation...
.. Get center of mass of the centerline/segmentation...
.. Computing physical coordinates of centerline/segmentation...
.. Smoothing algo = nurbs

Fitting centerline using B-spline approximation...
Test: # of control points = 5
Error on approximation = 42.06 mm
Test: # of control points = 6
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 7
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 8
Error on approximation = 0.79 mm
Test: # of control points = 9
Error on approximation = 0.5 mm
Test: # of control points = 10
Traceback (most recent call last):
  File "/Users/julien/code/sct/scripts/sct_process_segmentation.py", line 1231, in <module>
    main(sys.argv[1:])
  File "/Users/julien/code/sct/scripts/sct_process_segmentation.py", line 242, in main
    compute_csa(fname_segmentation, output_folder, overwrite, verbose, remove_temp_files, step, smoothing_param, slices, vert_lev, fname_vertebral_labeling, algo_fitting=param.algo_fitting, type_window=param.type_window, window_length=param.window_length, angle_correction=angle_correction, use_phys_coord=use_phys_coord)
  File "/Users/julien/code/sct/scripts/sct_process_segmentation.py", line 647, in compute_csa
    x_centerline_fit, y_centerline_fit, z_centerline, x_centerline_deriv, y_centerline_deriv, z_centerline_deriv = smooth_centerline('segmentation_RPI.nii.gz', algo_fitting=algo_fitting, type_window=type_window, window_length=window_length, nurbs_pts_number=3000, phys_coordinates=True, verbose=verbose, all_slices=False)
  File "/Users/julien/code/sct/scripts/sct_straighten_spinalcord.py", line 157, in smooth_centerline
    z_centerline_deriv, mse = b_spline_nurbs(x_centerline, y_centerline, z_centerline, nbControl=None, path_qc=curdir, point_number=nurbs_pts_number, verbose=verbose, all_slices=all_slices)
  File "/Users/julien/code/sct/scripts/msct_smooth.py", line 298, in b_spline_nurbs
    nurbs = NURBS(degree, point_number, data, False, nbControl, verbose, all_slices=all_slices, twodim=twodim)
  File "/Users/julien/code/sct/scripts/msct_nurbs.py", line 152, in __init__
    self.pointsControle = self.reconstructGlobalApproximation(P_x, P_y, P_z, self.degre, self.nbControle, w)
  File "/Users/julien/code/sct/scripts/msct_nurbs.py", line 645, in reconstructGlobalApproximation
    P_xb = (R.T * W * R).I * Tx.T
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/numpy/matrixlib/defmatrix.py", line 972, in getI
    return asmatrix(func(self))
  File "/Users/julien/code/sct/python/lib/python2.7/site-packages/scipy/linalg/basic.py", line 687, in inv
    raise LinAlgError("singular matrix")
numpy.linalg.linalg.LinAlgError: singular matrix
~~~

### LINKED ISSUE #2031: Error: singular matrix during straightening
### Steps to Reproduce

Data: 
[data.zip](https://github.com/neuropoly/spinalcordtoolbox/files/2429310/data.zip)

Syntax + output:
~~~
--
Spinal Cord Toolbox (version: dev)
Running /usr/local/sct/scripts/sct_smooth_spinalcord.py -i volume.nii -s centerline_optic.nii.gz

Check input arguments...
  Volume to smooth .................. volume.nii
  Centerline ........................ centerline_optic.nii.gz
  Sigma (mm) ........................ 3
  Verbose ........................... 1

Create temporary folder (/tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK)...

Copying input data to tmp folder and convert to nii...
cp volume.nii /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/anat.nii
cp centerline_optic.nii.gz /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/centerline.nii.gz
sct_convert -i anat.nii -o anat.nii
WARNING: File anat.nii already exists. Will overwrite it.
sct_convert -i centerline.nii.gz -o centerline.nii

Orient input volume to RPI orientation...

Orient centerline to RPI orientation...

Straighten the spinal cord using centerline/segmentation...
sct_straighten_spinalcord -i /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/anat_rpi.nii -o anat_rpi_straight.nii -s /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/centerline_rpi.nii -x spline # in /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK
Traceback (most recent call last):
  File "/usr/local/sct/scripts/sct_smooth_spinalcord.py", line 217, in <module>
    main()
  File "/usr/local/sct/scripts/sct_smooth_spinalcord.py", line 169, in main
    sct.run(['sct_straighten_spinalcord', '-i', fname_anat_rpi, '-o', 'anat_rpi_straight.nii', '-s', fname_centerline_rpi, '-x', 'spline'], verbose)
  File "/usr/local/sct/scripts/sct_utils.py", line 476, in run
    raise RunError(output)
sct_utils.RunError: Running /usr/local/sct/scripts/sct_straighten_spinalcord.py -i /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/anat_rpi.nii -o anat_rpi_straight.nii -s /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/centerline_rpi.nii -x spline
/usr/local/sct

Check input arguments:
Input volume ...................... /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/anat_rpi.nii
Centerline ........................ /tmp/sct-20180927171734.121997-smooth_spinalcord-kN6GNK/centerline_rpi.nii
Final interpolation ............... spline
Verbose ........................... 1


Create temporary folder (/tmp/sct-20180927171734.831909-straighten_spinalcord-h_I4Vn)...

Copy files to tmp folder...

Orient centerline to RPI orientation...

Get dimensions...
.. matrix size: 39 x 48 x 34
.. voxel size:  1.45833mm x 1.45833mm x 4.24999mm

Smooth centerline/segmentation...
.. Get center of mass of the centerline/segmentation...
.. Computing physical coordinates of centerline/segmentation...
.. Smoothing algo = nurbs


Fitting centerline using B-spline approximation...
Test: # of control points = 5
Error on approximation = 7.78 mm
Test: # of control points = 6
Error on approximation = 0.9 mm
Test: # of control points = 7
Error on approximation = 0.91 mm
Test: # of control points = 8
Error on approximation = 0.68 mm
Test: # of control points = 9
Error on approximation = 0.56 mm
Test: # of control points = 10
Error on approximation = 0.42 mm
Test: # of control points = 11
Error on approximation = 0.38 mm
Test: # of control points = 12
Error on approximation = 0.34 mm
Test: # of control points = 13
Error on approximation = 0.33 mm
Test: # of control points = 14
Error on approximation = 0.3 mm
Test: # of control points = 15
Error on approximation = 0.3 mm
Test: # of control points = 16
Error on approximation = 0.29 mm
Test: # of control points = 17
Error on approximation = 0.29 mm
Test: # of control points = 18
Error on approximation = 0.31 mm
Test: # of control points = 19
Error on approximation = 0.35 mm
Test: # of control points = 20
Error on approximation = 0.41 mm
Test: # of control points = 21
Error on approximation = 0.47 mm
Test: # of control points = 22
Error on approximation = 0.58 mm
Test: # of control points = 23
Error on approximation = 0.59 mm
Test: # of control points = 24
Error on approximation = 0.57 mm
Test: # of control points = 25
Error on approximation = 0.64 mm
Test: # of control points = 26
Error on approximation = 0.7 mm
Test: # of control points = 27
Error on approximation = 0.79 mm
Test: # of control points = 28
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 29
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 30
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 31
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 32
WARNING: NURBS instability -> wrong reconstruction
Test: # of control points = 33
WARNING: Exception during Straightening:
Error on line 370
singular matrix
Traceback (most recent call last):
File "/usr/local/sct/scripts/sct_straighten_spinalcord.py", line 1090, in <module>
main()
File "/usr/local/sct/scripts/sct_straighten_spinalcord.py", line 1074, in main
fname_straight = sc_straight.straighten()
File "/usr/local/sct/scripts/sct_straighten_spinalcord.py", line 370, in straighten
centerline = Centerline(*smooth_centerline('centerline_rpi.nii.gz', algo_fitting=algo_fitting, type_window=type_window, window_length=window_length, verbose=verbose, nurbs_pts_number=number_of_points, all_slices=False, phys_coordinates=True, remove_outliers=True))
File "/usr/local/sct/scripts/sct_straighten_spinalcord.py", line 160, in smooth_centerline
z_centerline_deriv, mse = b_spline_nurbs(x_centerline, y_centerline, z_centerline, nbControl=None, path_qc=curdir, point_number=nurbs_pts_number, verbose=verbose, all_slices=all_slices)
File "/usr/local/sct/scripts/msct_smooth.py", line 238, in b_spline_nurbs
nurbs = NURBS(degree, point_number, data, False, nbControl, verbose, all_slices=all_slices, twodim=twodim)
File "/usr/local/sct/scripts/msct_nurbs.py", line 156, in __init__
self.pointsControle = self.reconstructGlobalApproximation(P_x, P_y, P_z, self.degre, self.nbControle, w)
File "/usr/local/sct/scripts/msct_nurbs.py", line 649, in reconstructGlobalApproximation
P_xb = (R.T * W * R).I * Tx.T
File "/usr/local/sct/python/lib/python2.7/site-packages/numpy/matrixlib/defmatrix.py", line 972, in getI
return asmatrix(func(self))
File "/usr/local/sct/python/lib/python2.7/site-packages/scipy/linalg/basic.py", line 687, in inv
raise LinAlgError("singular matrix")
numpy.linalg.linalg.LinAlgError: singular matrix
Sentry is attempting to send 1 pending error messages
Waiting up to 10 seconds
Press Ctrl-C to quit
Note: you can opt out of Sentry reporting by editing the file ${SCT_DIR}/bin/sct_launcher and delete the line starting with "export SENTRY_DSN"
Total processing time: 0 min 9 s
Sentry is attempting to send 1 pending error messages
Waiting up to 10 seconds
Press Ctrl-C to quit
Note: you can opt out of Sentry reporting by editing the file ${SCT_DIR}/bin/sct_launcher and delete the line starting with "export SENTRY_DSN"
Total processing time: 0 min 11 s
~~~

### LINKED ISSUE #2038: NURBS too sensitive to number of input points
NURBS is used in key functions, including `sct_straighten_spinalcord` (used for template registration, smoothing, etc.), `sct_propseg`, `sct_deepseg_sc`. There are a couple of issues that arose, in relation to the sensitivity of NURBS to the number of control points, the gap and the number of input points:
- https://github.com/neuropoly/spinalcordtoolbox/issues/1863
- https://github.com/neuropoly/spinalcordtoolbox/issues/2031

It would be nice to re-think the method, as suggested [here](https://github.com/neuropoly/spinalcordtoolbox/issues/2031#issuecomment-425689939), to increase the robustness of the method. 

Some more details about where nurbs is called:
- `sct_get_centerline` --> `sct_process_segmentation.extract_centerline`--> `sct_smooth_spinalcord.smooth_centerline` --> `msct_smooth.b_spline_nurbs` --> `msct_nurbs.NURBS`

### PR DESCRIPTION
**Pull request by Nicolas-Pinon**
This PR attends to fix #2038 (which fixes #1863 and fixes #2031 ) by intercepting the exception caused by a singular matrix. That way the testing of the number of control points can continue and the NURBS fitting can be done even if a singular matrix is encountered.
Thus, the hanning filtering will not be necessary anymore (if the number of slices if sufficient). Moving away from the hanning filtering is something @benjamindeleener suggested when discussing this problem.
However, no change in the method has been done, should I search for methods to improve the NURBS fitting ?

### CODE REVIEW

File: scripts/msct_smooth.py [hunk 1]
```diff
@@ -270,7 +270,7 @@ def b_spline_nurbs(x, y, z, fname_centerline=None, degree=3, point_number=3000,
             PC_z = [p[2] for p in PC]
 
         import matplotlib
-        matplotlib.use('Agg')  # prevent display figure
+        # matplotlib.use('Agg')  # prevent display figure
```
Comment (jcohenadad) [inline scripts/msct_smooth.py]: any reason for this change? (this does not seem related to the purpose of the PR)

Comment (Nicolas-Pinon) [reply scripts/msct_smooth.py]: That was for testing, my bad again, i'll discard this change

**Comment by Nicolas-Pinon** [comment]
> if this PR is supposed to also fix #1863 and #2031, i would explicitly mention it (so the issues can be closed upon PR merging)
done

**Comment by Nicolas-Pinon** [comment]
Agreed, new issue referenced as #2043

**Comment by Nicolas-Pinon** [comment]
Agreed, new issue referenced as #2043
